TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php

```diff
@@ -1476,4 +1530,69 @@ protected function addIdentifierConstraintToQueryBuilder(QueryBuilder $queryBuil
         $queryBuilder->andWhere('n.identifier = :identifier')
             ->setParameter('identifier', $identifier);
     }
+
+    /**
+     * @param array $nodeDataObjects
+     * @param Workspace $workspace
+     * @param array $dimensions
+     * @return array
+     */
+    protected function filterNodeDataByBestMatchInContext(array $nodeDataObjects, Workspace $workspace, array $dimensions)
+    {
+        $workspaces = $this->collectWorkspaceAndAllBaseWorkspaces($workspace);
+        $notPersistedNodes = [];
+        $nodeIdentifier = [];
+
+        /** @var NodeData $nodeData */
+        foreach ($nodeDataObjects as $nodeData) {
+            $nodeIdentifier[] = $nodeData->getIdentifier();
+            while ($workspace !== null) {
+                /** @var $node NodeData */
+                foreach ($this->addedNodes as $node) {
+                    if ($node->getIdentifier() === $nodeData->getIdentifier() && $node->matchesWorkspaceAndDimensions($workspace, $dimensions) && $node->isInternal() === false) {
+                        $notPersistedNodes[] = $node;
+                    }
+                }
+
+                $workspace = $workspace->getBaseWorkspace();
+            }
+        }
+
+        $queryBuilder = $this->createQueryBuilder($workspaces);
+        if ($dimensions !== null) {
+            $this->addDimensionJoinConstraintsToQueryBuilder($queryBuilder, $dimensions);
+        } else {
+            $dimensions = [];
+        }
+        $queryBuilder->andWhere('n.movedTo IS NULL OR n.removed = FALSE');
+        $queryBuilder->andWhere('n.identifier IN (:identifier)')
+            ->setParameter('identifier', $nodeIdentifier);
+        $query = $queryBuilder->getQuery();
+        $nodes = $query->getResult();
+        $foundNodes = array_merge($nodes, $notPersistedNodes);
+        $foundNodes = $this->reduceNodeVariantsByWorkspacesAndDimensions($foundNodes, $workspaces, $dimensions);
+        $foundNodes = $this->filterRemovedNodes($foundNodes, false);
+
+        /** @var NodeData $nodeData */
+        return array_filter($nodeDataObjects, function (NodeData $nodeData) use ($foundNodes) {
+            return (isset($foundNodes[$nodeData->getIdentifier()]) && $foundNodes[$nodeData->getIdentifier()] === $nodeData);
+        });
+    }
+
+    /**
+     * Returns an array that contains the given workspace and all base (parent) workspaces of it.
+     *
+     * @param Workspace $workspace
+     * @return array
+     */
+    protected function collectWorkspaceAndAllBaseWorkspaces(Workspace $workspace)
```

Comment:
Yes I know, I didn't want to make this into a big refactoring as it's a bugfix in lower versions. I think I didn something similar for 2.2 or master already